### ISSUE TITLE
Add a Uri.set counterpart to Uri.make

### ISSUE THREAD

**Issue by hcarty** [mirage/ocaml-uri#96]
It would be useful to have a `Uri.set` function (maybe with a better name?) with all of the sample optional arguments as `Uri.make` which takes an existing URI and replaces whichever values are supplied by the optional arguments.

```ocaml
val make : ?scheme:string -> ?userinfo:string -> ?host:string ->
  ?port:int -> ?path:string -> ?query:(string * string list) list ->
  ?fragment:string -> unit -> t

val set : ?scheme:string option -> ?userinfo:string option -> ?host:string option ->
  ?port:int option -> ?path:string option -> ?query:(string * string list) list option ->
  ?fragment:string option -> t -> t
```

For each optional argument provided, the result would be equivalent to stacking calls to the corresponding `Uri.with_*` function.  If all of the optional arguments were elided then the given URI would be returned as-is.

**Comment by rgrinberg** [comment]
How would we be able to unset some fields that _are_ options.

**Comment by hcarty** [comment]
If I understand the question correctly: `Uri.set ~scheme:None (Uri.of_string "scheme://host/path")` would result in the URI `//host/path`

**Comment by dsheets** [comment]
This is very similar to #89 and a well-reasoned PR would be welcome.

**Comment by rgrinberg** [comment]
>If I understand the question correctly: `Uri.set ~scheme:None (Uri.of_string "scheme://host/path")` would result in the URI `//host/path`

OK, then I misunderstood. I thought fields that weren't set wouldn't be updated at all.

**Comment by agarwal** [comment]
> I thought fields that weren't set wouldn't be updated at all.

That's not wrong if by "not set" you mean "not given". @hcarty's proposal correctly provides a double option. Outer `None` means don't set the value, but `Some None` means set it to `None.

**Comment by rgrinberg** [comment]
On 02/04, Ashish Agarwal wrote:
>That's not wrong if by "not set" you mean "not given". @hcarty's proposal correctly provides a double option. Outer `None` means don't set the value, but `Some None` means set it to `None.

Thanks, I missed that double option bit.

**Comment by hcarty** [comment]
I think this can be closed now that #97 has been merged.
